<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to the SpatialFeatureExperiment class • SpatialFeatureExperiment</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the SpatialFeatureExperiment class">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialFeatureExperiment</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.7.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/SFE.html">Introduction to the SpatialFeatureExperiment class</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/SpatialFeatureExperiment/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to the SpatialFeatureExperiment class</h1>
                        <h4 data-toc-skip class="author">Lambda Moses,
Lior Pachter</h4>
            
            <h4 data-toc-skip class="date">Jul 22, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/SpatialFeatureExperiment/blob/documentation-devel/vignettes/SFE.Rmd" class="external-link"><code>vignettes/SFE.Rmd</code></a></small>
      <div class="d-none name"><code>SFE.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>This package can be installed from Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="class-structure">Class structure<a class="anchor" aria-label="anchor" href="#class-structure"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="http://adv-r.had.co.nz/S4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). <code>SpatialFeatureExperiment</code> incorporates geometries and
geometry operations with the <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package. Examples of supported geometries are Visium spots represented
with polygons corresponding to their size, cell or nuclei segmentation
polygons, tissue boundary polygons, pathologist annotation of
histological regions, and transcript spots of genes. Using
<code>sf</code>, <code>SpatialFeatureExperiment</code> leverages the
GEOS C++ libraries underlying <code>sf</code> for geometry operations,
including algorithms for for determining whether geometries intersect,
finding intersection geometries, buffering geometries with margins, etc.
A schematic of the SFE object is shown below:</p>
<div class="figure">
<img src="sfe_schematics.png" alt="SpatialFeatureExperiment expands on SpatialExperiment by adding column, row, and annotation geometries and spatial graphs. This is explained in detail in the following paragraphs." width="100%"><p class="caption">
Schematics of the SFE object
</p>
</div>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. A potential use case is to
store transcript spots for each gene in smFISH or in situ sequencing
based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
<li>
<code>localResults</code> are similar to <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>, but stores results from
univariate and bivariate local spatial analysis results, such as from <a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link"><code>localmoran</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">Getis-Ord
Gi*</a>, and <a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">local
spatial heteroscedasticity (LOSH)</a>. Unlike in
<code>reducedDims</code>, for each type of results (type is the type of
analysis such as Getis-Ord Gi*), each feature (e.g. gene) or pair of
features for which the analysis is performed has its own results. The
local spatial analyses can also be performed for attributes of
<code>colGeometries</code> and <code>annotGeometries</code> in addition
to gene expression and <code>colData</code>. Results of multivariate
spatial analysis such as <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html#multispati-analysis" class="external-link">MULTISPATI
PCA</a> can be stored in <code>reducedDims</code>.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SpatialFeatureExperiment'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     scale</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,</span></span>
<span><span class="co">#&gt;     tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.10.2, GDAL 3.4.1, PROJ 8.2.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Matrix'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/RBioFormats" class="external-link">RBioFormats</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; BioFormats library version 7.3.0</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example dataset</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 77 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(77): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCTTTGGTCGCGACG</span></span>
<span><span class="co">#&gt;   TTGATGTGTAGTCCCG</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="geometries">Geometries<a class="anchor" aria-label="anchor" href="#geometries"></a>
</h3>
<p>User interfaces to get or set the geometries and spatial graphs
emulate those of <code>reducedDims</code> and <code>row/colPairs</code>
in <code>SingleCellExperiment</code>. Column and row geometries also
emulate <code>reducedDims</code> in internal implementation, while
annotation geometries and spatial graphs differ.</p>
<div class="section level4">
<h4 id="column-and-row">Column and row<a class="anchor" aria-label="anchor" href="#column-and-row"></a>
</h4>
<p>Column and row geometries can be get or set with the
<code><a href="../reference/dimGeometries.html">dimGeometries()</a></code> or <code><a href="../reference/dimGeometries.html">dimGeometry()</a></code> function. The
<code>MARGIN</code> argument is as in the <code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code> function:
<code>MARGIN = 1</code> means row, and <code>MARGIN = 2</code> means
column.</p>
<p><code><a href="../reference/dimGeometries.html">dimGeometry()</a></code> gets or sets one particular geometry by
name of index.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get Visium spot polygons</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimGeometries.html">dimGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"spotPoly"</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 77 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 15000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGCGACTCCTG POLYGON ((5501.981 14118.62...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGAGTACGGATGCCC POLYGON ((5084.397 13395.63...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="va">spots</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="../reference/dimGeometries.html">dimGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"foobar"</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p><code><a href="../reference/dimGeometries.html">dimGeometries()</a></code> gets or sets all geometry of the given
margin.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter, all geometries of one margin</span></span>
<span><span class="op">(</span><span class="va">cgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimGeometries.html">dimGeometries</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 2</span></span>
<span><span class="co">#&gt; names(2): spotPoly foobar</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter, all geometries</span></span>
<span><span class="fu"><a href="../reference/dimGeometries.html">dimGeometries</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cgs</span></span></code></pre></div>
<p><code><a href="../reference/dimGeometries.html">dimGeometryNames()</a></code> gets or sets the names of the
geometries</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">cg_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimGeometries.html">dimGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "spotPoly" "foobar"</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="../reference/dimGeometries.html">dimGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cg_names</span></span></code></pre></div>
<p><code>colGeometry(sfe, "spotPoly")</code>,
<code>colGeometries(sfe)</code>, and <code>colGeometryNames(sfe)</code>
are shorthands for
<code>dimGeometry(sfe, "spotPoly", MARGIN = 2)</code>,
<code>dimGeometries(sfe, MARGIN = 2)</code>, and
<code>dimGeometryNames(sfe, MARGIN = 2)</code> respectively. Similarly,
<code>rowGeometr*(sfe, ...)</code> is a shorthand of
<code>dimGeometr*(sfe, ..., MARGIN = 1)</code>.</p>
<p>There are shorthands for some specific column or row geometries. For
example, <code>spotPoly(sfe)</code> is equivalent to
<code>colGeometry(sfe, "spotPoly")</code> for Visium spot polygons, and
<code>txSpots(sfe)</code> is equivalent to
<code>rowGeometry(sfe, "txSpots")</code> for transcript spots in single
molecule technologies.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 77 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 15000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGCGACTCCTG POLYGON ((5501.981 14118.62...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGAGTACGGATGCCC POLYGON ((5084.397 13395.63...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h4>
<p>Annotation geometries can be get or set with
<code><a href="../reference/annotGeometries.html">annotGeometries()</a></code> or <code><a href="../reference/annotGeometries.html">annotGeometry()</a></code>. In
column or row geometries, the number of rows of the <code>sf</code> data
frame (i.e. the number of geometries in the data frame) is constrained
by the number of rows or columns of the gene count matrix respectively,
because just like <code>rowData</code> and <code>colData</code>, each
row of a <code>rowGeometry</code> or <code>colGeometry</code>
<code>sf</code> data frame must correspond to a row or column of the
gene count matrix respectively. In contrast, an
<code>annotGeometry</code> <code>sf</code> data frame can have any
dimension, not constrained by the dimension of the gene count matrix.
Similar to column and row geometries, annotation geometries have
<code><a href="../reference/annotGeometries.html">annotGeometry()</a></code>, <code><a href="../reference/annotGeometries.html">annotGeometries()</a></code>, and
<code><a href="../reference/annotGeometries.html">annotGeometryNames()</a></code> getters and setters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter, by name or index</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotGeometries.html">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 1 feature and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5094 ymin: 13000 xmax: 7000 ymax: 14969</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   ID                       geometry sample_id</span></span>
<span><span class="co">#&gt; 7  7 POLYGON ((5094 13000, 5095 ...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter, by name or index</span></span>
<span><span class="fu"><a href="../reference/annotGeometries.html">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get all annoation geometries as named list</span></span>
<span><span class="va">ags</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotGeometries.html">annotGeometries</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set all annotation geometries with a named list</span></span>
<span><span class="fu"><a href="../reference/annotGeometries.html">annotGeometries</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ags</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get names of annotation geometries</span></span>
<span><span class="op">(</span><span class="va">ag_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotGeometries.html">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set names</span></span>
<span><span class="fu"><a href="../reference/annotGeometries.html">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ag_names</span></span></code></pre></div>
<p>There are shorthands for specific annotation geometries. For example,
<code>tissueBoundary(sfe)</code> is equivalent to
<code>annotGeometry(sfe, "tissueBoundary")</code>.
<code><a href="../reference/colGeometries.html">cellSeg()</a></code> (cell segmentation) and <code><a href="../reference/colGeometries.html">nucSeg()</a></code>
(nuclei segmentation) would first query <code>colGeometries</code> (for
single cell, single molecule technologies, equivalent to
<code>colGeometry(sfe, "cellSeg")</code> or
<code>colGeometry(sfe, "nucSeg")</code>), and if not found, they will
query <code>annotGeometries</code> (for array capture and
microdissection technologies, equivalent to
<code>annotGeometry(sfe, "cellSeg")</code> or
<code>annotGeometry(sfe, "nucSeg")</code>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 1 feature and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5094 ymin: 13000 xmax: 7000 ymax: 14969</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   ID                       geometry sample_id</span></span>
<span><span class="co">#&gt; 7  7 POLYGON ((5094 13000, 5095 ...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="spatial-graphs">Spatial graphs<a class="anchor" aria-label="anchor" href="#spatial-graphs"></a>
</h3>
<p>Column, row, and annotation spatial graphs can be get or set with
<code><a href="../reference/spatialGraphs.html">spatialGraphs()</a></code> and <code><a href="../reference/spatialGraphs.html">spatialGraph()</a></code> functions.
Similar to <code>dimGeometr*</code> functions,
<code>spatialGraph*</code> functions have a <code>MARGIN</code>
argument. However, since internally, row and column geometries are
implemented very differently from annotation geometries, while row,
column, and annotation graphs are implemented the same way, for the
<code>spatialGraph*</code> functions, <code>MARGIN = 1</code> means
rows, <code>MARGIN = 2</code> means columns, and <code>MARGIN = 3</code>
means annotation. Similar to <code>dimGeometry*</code> functions, there
are <code>rowGraph*</code>, <code>colGraph*</code>, and
<code>annotGraph*</code> getter and setter functions for each
margin.</p>
<p>This package wraps functions in the <code>spdep</code> package to
find spatial neighborhood graphs. In this example, triangulation is used
to find the spatial graph; many other methods are also supported, such
as k nearest neighbors, distance based neighbors, and polygon
contiguity.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSpatialNeighbors-SpatialFeatureExperiment-method.html">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"tri2nb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 428 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.218755 </span></span>
<span><span class="co">#&gt; Average number of links: 5.558442 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0      S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 28.0096 309.4083</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">g</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph by name</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">spatialGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span>
<span><span class="co"># Or equivalently</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph by name</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialGraphs.html">spatialGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span>, MARGIN <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="co"># Or equivalently</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span><span class="op">)</span></span>
<span><span class="va">g</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 428 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.218755 </span></span>
<span><span class="co">#&gt; Average number of links: 5.558442 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0      S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 28.0096 309.4083</span></span></code></pre></div>
<p>For Visium, spatial neighborhood graph of the hexagonal grid can be
found with the known locations of the barcodes.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findVisiumGraph.html">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>All graphs of the SFE object, or if specified, of the margin of
interest, can be get or set with <code><a href="../reference/spatialGraphs.html">spatialGraphs()</a></code> and the
margin specific wrappers.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialGraphs.html">colGraphs</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; $col</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 428 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.218755 </span></span>
<span><span class="co">#&gt; Average number of links: 5.558442 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0      S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 28.0096 309.4083</span></span></code></pre></div>
<p>Similar to <code><a href="../reference/dimGeometries.html">dimGeometries()</a></code>, the graphs have
<code><a href="../reference/spatialGraphs.html">spatialGraphNames()</a></code> getter and setter and the margin
specific wrappers.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialGraphs.html">colGraphNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "graph1" "visium"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="multiple-samples">Multiple samples<a class="anchor" aria-label="anchor" href="#multiple-samples"></a>
</h3>
<p>Thus far, the example dataset used only has one sample. The
<code>SpatialExperiment</code> (SPE) object has a special column in
<code>colData</code> called <code>sample_id</code>, so data from
multiple tissue sections can coexist in the same SPE object for joint
dimension reduction and clustering while keeping the spatial coordinates
separate. It’s important to keep spatial coordinates of different tissue
sections separate because first, the coordinates would only make sense
within the same section, and second, the coordinates from different
sections can have overlapping numeric values.</p>
<p>SFE inherits from SPE, and with geometries and spatial graphs,
<code>sample_id</code> is even more important. The geometry and graph
getter and setter functions have a <code>sample_id</code> argument,
which is optional when only one sample is present in the SFE object.
This argument is mandatory if multiple samples are present, and can be a
character vector for multiple samples or “all” for all samples. Below
are examples of using the getters and setters for multiple samples.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct toy dataset with 2 samples</span></span>
<span><span class="va">sfe1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">sfe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe2</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"sample02"</span></span>
<span><span class="op">(</span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sfe1</span>, <span class="va">sfe2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 149 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(149): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCCTCGGACTAACCA</span></span>
<span><span class="co">#&gt;   TTCTGACCGGGCTCAA</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (GEOMETRY), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>Use the <code>sampleIDs</code> function to see the names of all
samples</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sampleIDs.html">sampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Vis5A"    "sample02"</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only get the geometries for the second sample</span></span>
<span><span class="op">(</span><span class="va">spots2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colGeometries.html">colGeometry</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 72 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6000 ymin: 7025.865 xmax: 8000 ymax: 9000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                  sample_id                       geometry</span></span>
<span><span class="co">#&gt; AACACACGCTCGCCGC  sample02 POLYGON ((6597.869 7842.575...</span></span>
<span><span class="co">#&gt; AACCGCTAAGGGATGC  sample02 POLYGON ((6724.811 9000, 67...</span></span>
<span><span class="co">#&gt; AACGCTGTTGCTGAAA  sample02 POLYGON ((6457.635 7118.991...</span></span>
<span><span class="co">#&gt; AACGGACGTACGTATA  sample02 POLYGON ((6737.064 8083.571...</span></span>
<span><span class="co">#&gt; AATAGAATCTGTTTCA  sample02 POLYGON ((7570.153 8564.368...</span></span>
<span><span class="co">#&gt; ACAAATCGCACCGAAT  sample02 POLYGON ((8000 7997.001, 79...</span></span>
<span><span class="co">#&gt; ACAATTGTGTCTCTTT  sample02 POLYGON ((6043.169 7843.77,...</span></span>
<span><span class="co">#&gt; ACAGGCTTGCCCGACT  sample02 POLYGON ((7428.88 7358.195,...</span></span>
<span><span class="co">#&gt; ACCAGTGCGGGAGACG  sample02 POLYGON ((6460.753 8566.757...</span></span>
<span><span class="co">#&gt; ACCCTCCCTTGCTATT  sample02 POLYGON ((7847.503 8563.771...</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only set the geometries for the second sample</span></span>
<span><span class="co"># Leaving geometries of the first sample intact</span></span>
<span><span class="fu"><a href="../reference/colGeometries.html">colGeometry</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots2</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph only for the second sample</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"foo"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/findSpatialNeighbors-SpatialFeatureExperiment-method.html">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph only for the second sample</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"foo"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 406 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.83179 </span></span>
<span><span class="co">#&gt; Average number of links: 5.638889 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 25.82104 289.8299</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph of the same name for both samples</span></span>
<span><span class="co"># The graphs are computed separately for each sample</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span>, name <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/findVisiumGraph.html">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get multiple graphs of the same name</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span>, name <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Vis5A</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 394 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 6.645303 </span></span>
<span><span class="co">#&gt; Average number of links: 5.116883 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 31.68056 311.7544</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 366 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.060185 </span></span>
<span><span class="co">#&gt; Average number of links: 5.083333 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 29.83889 291.5833</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or just all graphs of the margin</span></span>
<span><span class="fu"><a href="../reference/spatialGraphs.html">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Vis5A</span></span>
<span><span class="co">#&gt; $Vis5A$visium</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 394 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 6.645303 </span></span>
<span><span class="co">#&gt; Average number of links: 5.116883 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 31.68056 311.7544</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02</span></span>
<span><span class="co">#&gt; $sample02$foo</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 406 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.83179 </span></span>
<span><span class="co">#&gt; Average number of links: 5.638889 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 25.82104 289.8299</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02$visium</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 366 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.060185 </span></span>
<span><span class="co">#&gt; Average number of links: 5.083333 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 29.83889 291.5833</span></span></code></pre></div>
<p>Sample IDs can also be changed, with the
<code><a href="../reference/changeSampleIDs.html">changeSampleIDs()</a></code> function, with a named vector whose names
are the old names and values are the new names.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/changeSampleIDs.html">changeSampleIDs</a></span><span class="op">(</span><span class="va">sfe</span>, replacement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Vis5A <span class="op">=</span> <span class="st">"foo"</span>, sample02 <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sfe_combined</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 77 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(77): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCTTTGGTCGCGACG</span></span>
<span><span class="co">#&gt;   TTGATGTGTAGTCCCG</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON), foobar (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; foo: col: graph1, visium</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="object-construction">Object construction<a class="anchor" aria-label="anchor" href="#object-construction"></a>
</h2>
<div class="section level3">
<h3 id="from-scratch">From scratch<a class="anchor" aria-label="anchor" href="#from-scratch"></a>
</h3>
<p>An SFE object can be constructed from scratch with the assay matrices
and metadata. In this toy example, <code>dgCMatrix</code> is used, but
since SFE inherits from SingleCellExperiment (SCE), other types of
arrays supported by SCE such as delayed arrays should also work.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>That should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the toy example can be
converted into spot polygons with the <code>spotDiameter</code>
argument. Spot diameter in pixels in full resolution image can be found
in the <code>scalefactors_json.json</code> file in Space Ranger
output.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialFeatureExperiment.html">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>When <code>colData</code> contains columns for the centroid
coordinates, the <code>spatialCoordsNames</code> argument specifies
which columns in <code>colData</code> are for the coordinates, in the
same order as x, y, and z (if applicable). If the coordinates are not in
<code>colData</code>, they can be specified separately in the
<code>spatialCoords</code> argument:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialFeatureExperiment.html">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, </span>
<span>                                 spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                 spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>More geometries and spatial graphs can be added after calling the
constructor.</p>
<p>Geometries can also be supplied in the constructor.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert regular data frame with coordinates to sf data frame</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/df2sf.html">df2sf</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>, spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialFeatureExperiment.html">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colGeometries <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">cg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="space-ranger-output">Space Ranger output<a class="anchor" aria-label="anchor" href="#space-ranger-output"></a>
</h3>
<p>Space Ranger output can be read in a similar manner as in
<code>SpatialExperiment</code>; the returned SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix is read in, then a column graph called
<code>visium</code> will also be present, for the spatial neighborhood
graph of the Visium spots on tissue. The graph is not computed if all
spots are read in regardless of whether they are on tissue.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample01"</span>, <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span><span class="op">)</span></span></code></pre></div>
<p>Inside the <code>outs</code> directory:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "filtered_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>There should also be <code>raw_feature_bc_matrix</code> though this
toy example only has the filtered matrix.</p>
<p>Inside the matrix directory:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"filtered_feature_bc_matrix"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv" "features.tsv" "matrix.mtx"</span></span></code></pre></div>
<p>Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aligned_fiducials.jpg"              "barcode_fluorescence_intensity.csv"</span></span>
<span><span class="co">#&gt; [3] "detected_tissue_image.jpg"          "scalefactors_json.json"            </span></span>
<span><span class="co">#&gt; [5] "spatial_enrichment.csv"             "tissue_hires_image.png"            </span></span>
<span><span class="co">#&gt; [7] "tissue_lowres_image.png"            "tissue_positions.csv"</span></span></code></pre></div>
<p>Not all Visium datasets have all the files here. The
<code>barcode_fluorescence_intensity.csv</code> file is only present for
datasets with fluorescent imaging rather than bright field H&amp;E.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10xVisiumSFE.html">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                          data <span class="op">=</span> <span class="st">"filtered"</span>, images <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample02</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample02</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(14): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample02</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample02</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(10): in_tissue array_row ... channel3_mean channel3_stdev</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: col: visium</span></span>
<span><span class="co">#&gt; sample02: col: visium</span></span></code></pre></div>
<p>The <code>barcode_fluorescence_intensity.csv</code> file is read into
<code>colData</code>. The <code>spatial_enrichment.csv</code> file
contains Moran’s I and its p-values for each gene; it is read into
<code>rowData</code>.</p>
<p>Instead of pixels in the full resolution image, the Visium data can
be read so the units are microns. Full resolution pixels is related to
microns by the spacing between spots, which is known to be 100 microns.
The unit can be set in the <code>unit</code> argument; for now only
“micron” and “full_res_image_pixel” are supported for Visium:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10xVisiumSFE.html">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                          data <span class="op">=</span> <span class="st">"filtered"</span>, images <span class="op">=</span> <span class="st">"hires"</span>, unit <span class="op">=</span> <span class="st">"micron"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting pixels to microns</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample02</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting pixels to microns</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample02</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(14): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample02</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample02</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(10): in_tissue array_row ... channel3_mean channel3_stdev</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: col: visium</span></span>
<span><span class="co">#&gt; sample02: col: visium</span></span></code></pre></div>
<p>The unit of the SFE object can be checked:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unit-SpatialFeatureExperiment-method.html">unit</a></span><span class="op">(</span><span class="va">sfe3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "micron"</span></span></code></pre></div>
<p>At present, this is merely a string and SFE doesn’t perform unit
conversion.</p>
<p>Unlike in <code>SpatialExperiment</code>, SFE reads the images as
<code><a href="https://rspatial.github.io/terra/reference/SpatRaster-class.html" class="external-link">terra::SpatRaster</a></code> objects, so the images are not loaded
into memory unless necessary. Also, with <code>terra</code>, if a larger
image is associated with the SFE object, it will not be fully loaded
into memory when plotted; rather, it’s downsampled.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/imgRaster.html">imgRaster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SpatRaster"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "terra"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vizgen-merfish-output">Vizgen MERFISH output<a class="anchor" aria-label="anchor" href="#vizgen-merfish-output"></a>
</h3>
<p>The commercialized MERFISH from Vizgen has a standard output format,
that can be read into SFE with <code><a href="../reference/readVizgen.html">readVizgen()</a></code>. Because the
cell segmentation from each field of view (FOV) has a separate HDF5 file
and a MERFISH dataset can have hundreds of FOVs, we strongly recommend
reading the MERFISH output on a server with a large number of CPU cores.
Alternatively, some but not all MERFISH datasets store cell segmentation
in a <code>parquet</code> file, which can be more easily read into R.
This requires the installation of <code>arrow</code>. Here we read a toy
dataset which is the first FOV from a real dataset:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/VizgenOutput.html" class="external-link">VizgenOutput</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"vizgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpmUkX4b/vizgen/vizgen_cellbound</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries"          "cell_boundaries.parquet" </span></span>
<span><span class="co">#&gt; [3] "cell_by_gene.csv"         "cell_metadata.csv"       </span></span>
<span><span class="co">#&gt; [5] "detected_transcripts.csv" "images"</span></span></code></pre></div>
<p>The optional <code>add_molecules</code> argument can be set to
<code>TRUE</code> to read in the transcript spots</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVizgen.html">readVizgen</a></span><span class="op">(</span><span class="va">dir_use</span>, z <span class="op">=</span> <span class="fl">3L</span>, image <span class="op">=</span> <span class="st">"PolyT"</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 1 `.parquet` files exist:</span></span>
<span><span class="co">#&gt; /tmp/RtmpmUkX4b/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; using -&gt; /tmp/RtmpmUkX4b/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file</span></span>
<span><span class="co">#&gt; Removing 35 cells with area less than 15</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering geometries to match 1023 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 88 1023 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(88): CD4 TLL1 ... Blank-38 Blank-39</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1023): 112824700230101267 112824700230101269 ...</span></span>
<span><span class="co">#&gt;   112824700330100848 112824700330100920</span></span>
<span><span class="co">#&gt; colData names(11): fov volume ... solidity sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The unit is always in microns. To make it easier and faster to read
the data next time, the processed cell segmentation geometries and
transcript spots are written to the same directory where the data
resides:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries"              "cell_boundaries.parquet"     </span></span>
<span><span class="co">#&gt; [3] "cell_by_gene.csv"             "cell_metadata.csv"           </span></span>
<span><span class="co">#&gt; [5] "detected_transcripts.csv"     "detected_transcripts.parquet"</span></span>
<span><span class="co">#&gt; [7] "images"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="x-xenium-output">10X Xenium output<a class="anchor" aria-label="anchor" href="#x-xenium-output"></a>
</h3>
<p>SFE supports reading the output from Xenium Onboarding Analysis (XOA)
v1 and v2 with the function <code><a href="../reference/readXenium.html">readXenium()</a></code>. Especially for
XOA v2, <code>arrow</code> is strongly recommended. The cell and nuclei
polygon vertices and transcript spot coordinates are in
<code>parquet</code> files Similar to <code><a href="../reference/readVizgen.html">readVizgen()</a></code>,
<code><a href="../reference/readXenium.html">readXenium()</a></code> makes <code>sf</code> data frames from the
vertices and transcript spots and saves them as GeoParquet files.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpmUkX4b/xenium/xenium2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cell_boundaries.csv.gz"     "cell_boundaries.parquet"   </span></span>
<span><span class="co">#&gt;  [3] "cell_feature_matrix.h5"     "cells.csv.gz"              </span></span>
<span><span class="co">#&gt;  [5] "cells.parquet"              "experiment.xenium"         </span></span>
<span><span class="co">#&gt;  [7] "morphology_focus"           "nucleus_boundaries.csv.gz" </span></span>
<span><span class="co">#&gt;  [9] "nucleus_boundaries.parquet" "transcripts.csv.gz"        </span></span>
<span><span class="co">#&gt; [11] "transcripts.parquet"</span></span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RBioFormats issue: https://github.com/aoles/RBioFormats/issues/42</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readXenium.html">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making POLYGON cell geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving geometries to parquet files</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readXenium.html">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Preprocessed sf segmentations found</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 398 6272 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(398): ABCC11 ACE2 ... UnassignedCodeword_0488</span></span>
<span><span class="co">#&gt;   UnassignedCodeword_0497</span></span>
<span><span class="co">#&gt; rowData names(3): ID Symbol Type</span></span>
<span><span class="co">#&gt; colnames(6272): abclkehb-1 abcnopgp-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">#&gt; colData names(9): transcript_counts control_probe_counts ...</span></span>
<span><span class="co">#&gt;   nucleus_area sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON), nucSeg (MULTIPOLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cell_boundaries_sf.parquet"    "cell_boundaries.csv.gz"       </span></span>
<span><span class="co">#&gt;  [3] "cell_boundaries.parquet"       "cell_feature_matrix.h5"       </span></span>
<span><span class="co">#&gt;  [5] "cells.csv.gz"                  "cells.parquet"                </span></span>
<span><span class="co">#&gt;  [7] "experiment.xenium"             "morphology_focus"             </span></span>
<span><span class="co">#&gt;  [9] "nucleus_boundaries_sf.parquet" "nucleus_boundaries.csv.gz"    </span></span>
<span><span class="co">#&gt; [11] "nucleus_boundaries.parquet"    "transcripts.csv.gz"           </span></span>
<span><span class="co">#&gt; [13] "transcripts.parquet"           "tx_spots.parquet"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="nanostring-cosmx-output">Nanostring CosMX output<a class="anchor" aria-label="anchor" href="#nanostring-cosmx-output"></a>
</h3>
<p>This is similar to <code><a href="../reference/readVizgen.html">readVizgen()</a></code> and
<code><a href="../reference/readXenium.html">readXenium()</a></code>, except that the output doesn’t come with
images.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/CosMXOutput.html" class="external-link">CosMXOutput</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"cosmx"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpmUkX4b/cosmx/cosmx</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Run5642_S3_Quarter_exprMat_file.csv" </span></span>
<span><span class="co">#&gt; [2] "Run5642_S3_Quarter_metadata_file.csv"</span></span>
<span><span class="co">#&gt; [3] "Run5642_S3_Quarter_tx_file.csv"      </span></span>
<span><span class="co">#&gt; [4] "Run5642_S3_Quarter-polygons.csv"</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_cosmx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readCosMX.html">readCosMX</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Constructing cell polygons</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 960 27 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(960): Chrna4 Slc6a1 ... NegPrb9 NegPrb10</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(27): 367_1 368_1 ... 581_1 583_1</span></span>
<span><span class="co">#&gt; colData names(19): fov cell_ID ... Max.DAPI sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : CenterX_global_px CenterY_global_px</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries_sf.parquet"          </span></span>
<span><span class="co">#&gt; [2] "Run5642_S3_Quarter_exprMat_file.csv" </span></span>
<span><span class="co">#&gt; [3] "Run5642_S3_Quarter_metadata_file.csv"</span></span>
<span><span class="co">#&gt; [4] "Run5642_S3_Quarter_tx_file.csv"      </span></span>
<span><span class="co">#&gt; [5] "Run5642_S3_Quarter-polygons.csv"     </span></span>
<span><span class="co">#&gt; [6] "tx_spots.parquet"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="other-technologies">Other technologies<a class="anchor" aria-label="anchor" href="#other-technologies"></a>
</h3>
<p>A read function for Visium HD is in progress. Contribution for Akoya,
Molecular Cartography, and Curio Seeker are welcome. See the <a href="https://github.com/pachterlab/SpatialFeatureExperiment/issues" class="external-link">issues</a>.</p>
</div>
<div class="section level3">
<h3 id="coercion-from-spatialexperiment">Coercion from <code>SpatialExperiment</code><a class="anchor" aria-label="anchor" href="#coercion-from-spatialexperiment"></a>
</h3>
<p>SPE objects can be coerced into SFE objects. If column geometries or
spot diameter are not specified, then a column geometry called
“centroids” will be created.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html" class="external-link">read10xVisium</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"filtered"</span>, </span>
<span>  images <span class="op">=</span> <span class="st">"hires"</span>, load <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For the coercion, column names must not be duplicate.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialFeatureExperiment-coercion.html">toSpatialFeatureExperiment</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit:</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>If images are present in the SPE object, they will be converted into
<code>SpatRaster</code> when the SPE object is converted into SFE.
Plotting functions in the <code>Voyager</code> package relies on
<code>SpatRaster</code> to plot the image behind the geometries.</p>
</div>
<div class="section level3">
<h3 id="coercion-from-seurat">Coercion from <code>Seurat</code><a class="anchor" aria-label="anchor" href="#coercion-from-seurat"></a>
</h3>
<p>Seurat objects canbe coerced into SFE objects though coercion from
SFE to Seurat is not yet implemented.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">obj_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_extdata</span>, <span class="st">"seu_vis_toy.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_conv_vis</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/SpatialFeatureExperiment-coercion.html">toSpatialFeatureExperiment</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">obj_vis</span>, </span>
<span>                             image_scalefactors <span class="op">=</span> <span class="st">"lowres"</span>,</span>
<span>                             unit <span class="op">=</span> <span class="st">"micron"</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Seurat Assays found: RNA</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; RNA -&gt; will be used as 'Main Experiment'</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Seurat spatial object found: VisiumV1</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 'full_res_image_pixel' units will be used -&gt;</span></span>
<span><span class="co">#&gt; ie 'imagerow' &amp; 'imagecol' without scaling factors</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; set `unit = 'micron'` to convert spot coordinates to micron space</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Generating `sf` geometries</span></span>
<span><span class="co">#&gt; Warning: Layer 'data' is empty</span></span>
<span><span class="co">#&gt; Warning: Layer 'scale.data' is empty</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Creating `SFE` object -&gt; sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting pixels to microns</span></span>
<span><span class="va">sfe_conv_vis</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 12 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ACPP KLK3 MSMB TGM4 TAGLN</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(12): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   CTTCCTGCATATTTAC-1 CAATATGTAGATTTAC-1</span></span>
<span><span class="co">#&gt; colData names(7): orig.ident nCount_RNA ... in_tissue sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : X Y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="operations">Operations<a class="anchor" aria-label="anchor" href="#operations"></a>
</h2>
<div class="section level3">
<h3 id="non-geometric">Non-geometric<a class="anchor" aria-label="anchor" href="#non-geometric"></a>
</h3>
<p>SFE objects can be concatenated with <code>cbind</code>, as was done
just now to create a toy example with 2 samples.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sfe1</span>, <span class="va">sfe2</span><span class="op">)</span></span></code></pre></div>
<p>The SFE object can also be subsetted like a matrix, like an SCE
object. More complexity arises when it comes to the spatial graphs. The
<code>drop</code> argument of the SFE method <code>[</code> determines
what to do with the spatial graphs. If <code>drop = TRUE</code>, then
all spatial graphs will be removed, since the graphs with nodes and
edges that have been removed are no longer valid. If
<code>drop = FALSE</code>, which is the default, then the spatial graphs
will be reconstructed with the remaining nodes after subsetting.
Reconstruction would only work when the original graphs were constructed
with <code>findSpatialNeighbors</code> or <code>findVisiumGraph</code>
in this package, which records the method and parameters used to
construct the graphs. If reconstruction fails, then a waning will be
issued and the graphs removed.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_subset</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; Node indices in the graphs are no longer valid after subsetting. Dropping all row and col graphs.</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 10 10 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(10): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000090031 ENSMUSG00000033740</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(10): AAATTACCTATCGATG AACATATCAACTGGTG ... ACGAGTACGGATGCCC</span></span>
<span><span class="co">#&gt;   ACGCTAGTGATACACT</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON), foobar (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Will give warning because graph reconstruction fails</span></span>
<span><span class="va">sfe_subset</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<p>If images are present, then they will be cropped to the bounding box
of the remaining geometries after subsetting.</p>
</div>
<div class="section level3">
<h3 id="geometric">Geometric<a class="anchor" aria-label="anchor" href="#geometric"></a>
</h3>
<p>Just like <code>sf</code> data frames, SFE objects can be subsetted
by a geometry and a predicate relating geometries. For example, if all
Visium spots were read into an SFE object regardless of whether they are
in tissue, and the <code>tissueBoundary</code> annotation geometry is
provided, then the tissue boundary geometry can be used to subset the
SFE object to obtain a new SFE object with only spots on tissue. Loupe
does not give the tissue boundary polygon; such polygon can be obtained
by thresholding the H&amp;E image and converting the mask into polygons
with OpenCV or the <code>terra</code> R package, or by manual annotation
in QuPath or LabKit (the latter needs to be converted into polygon).</p>
<div class="section level4">
<h4 id="crop">Crop<a class="anchor" aria-label="anchor" href="#crop"></a>
</h4>
<p>Use the <code>crop</code> function to directly get the subsetted SFE
object. When images are present, they are cropped by the bounding box of
the cropped geometries.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-70-1.png" width="700"></p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_in_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crop.html">crop</a></span><span class="op">(</span><span class="va">sfe</span>, y <span class="op">=</span> <span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p>Note that for large datasets with many geometries, cropping can take
a while to run.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-72-1.png" width="700"></p>
<p><code>crop</code> can also be used in the conventional sense of
cropping, i.e. specifying a bounding box.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_cropped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crop.html">crop</a></span><span class="op">(</span><span class="va">sfe</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5500</span>, xmax <span class="op">=</span> <span class="fl">6500</span>, ymin <span class="op">=</span> <span class="fl">13500</span>, ymax <span class="op">=</span> <span class="fl">14500</span><span class="op">)</span>,</span>
<span>                    colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>colGeometryName</code> is used to determine which columns
in the gene count matrix to keep. All geometries in the SFE object will
be subsetted so only portions intersecting <code>y</code> or the
bounding box are kept. Since the intersection operation can produce a
mixture of geometry types, such as intersection of two polygons
producing polygons, points, and lines, the geometry types of the
<code>sf</code> data frames after subsetting may be different from those
of the originals.</p>
<p>The cropping is done independently for each <code>sample_id</code>,
and only on <code>sample_id</code>s specified. Again,
<code>sample_id</code> is optional when there is only one sample in the
SFE object.</p>
<p>Geometry predicates and operations can also be performed to return
the results without subsetting an SFE object. For example, one may want
a logical vector indicating whether each Visium spot intersects the
tissue, or a numeric vector of how many nuclei there are in each Visium
spot. Or get the intersections between each Visium spot and nuclei.
Again, the geometry predicates and operations are performed
independently for each sample, and the <code>sample_id</code> argument
is optional when there is only one sample.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get logical vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotPred.html">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                                    annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                                    sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="co"># Get the number of nuclei per Visium spot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">n_nuclei</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotPred.html">annotNPred</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"spotPoly"</span>, annotGeometryName <span class="op">=</span> <span class="st">"nuclei"</span><span class="op">)</span></span>
<span><span class="co"># Get geometries of intersections of Visium spots and myofibers</span></span>
<span><span class="va">spot_intersections</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotOp.html">annotOp</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<p>Sometimes the spatial coordinates of different samples can take very
different values. The values can be made more comparable by moving all
tissues so the bottom left corner of the bounding box would be at the
origin, which would facilitate plotting and comparison across samples
with <code>geom_sf</code> and <code>facet_*</code>.</p>
<p>To find the bounding box of all geometries in each sample of an SFE
object:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/bbox-SpatialFeatureExperiment-method.html">bbox</a></span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  xmin  ymin  xmax  ymax </span></span>
<span><span class="co">#&gt;  5000 13000  7000 15000</span></span></code></pre></div>
<p>To move the coordinates:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_moved</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeEmptySpace.html">removeEmptySpace</a></span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span></code></pre></div>
<p>The original bounding box before moving is stored within the SFE
object, which can be read by <code>dimGeometry</code> setters so newly
added geometries can have coordinates moved as well; this behavior can
be turned off with the optional argument <code>translate = FALSE</code>
in <code>dimGeometry</code> setters.</p>
</div>
<div class="section level4">
<h4 id="transform">Transform<a class="anchor" aria-label="anchor" href="#transform"></a>
</h4>
<p>When images are present, they might need to be flipped to align with
the spots. <code>SpatialExperiment</code> implements methods to rotate
and mirror images, and SFE implements methods for SFE objects to
transpose and mirror images (<code><a href="https://rspatial.github.io/terra/reference/rotate.html" class="external-link">terra::rotate()</a></code> does NOT
rotate the image in the conventional sense – rather it changes the
longitudes and where the globe is cut to project to 2D just like cutting
a world map at the Atlantic vs. the Pacific).</p>
<p><code>SpatialExperiment</code> represents images with S4 classes
inheriting from the <code>VirtualSpatialImage</code> virtual class. To
be compatible with SPE, SFE uses <code>SpatRasterImage</code>, which is
a thin wrapper of <code>SpatRaster</code> inheriting from the virtual
class. Transformations can be applied to <code>SpatRasterImage</code>,
as well as SFE objects with sample and image IDs specified.</p>
<p>When an image is transposed, it is flipped about the line going from
top left to bottom right:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe3</span>, image_id <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/imgRaster.html">imgRaster</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-77-1.png" width="96"></p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/transposeImg.html">transposeImg</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/imgRaster.html">imgRaster</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-77-2.png" width="96"></p>
<p>Arguments for the SFE method of <code><a href="../reference/mirrorImg.html">mirrorImg()</a></code> differ from
those of the SPE method, to match <code><a href="https://rspatial.github.io/terra/reference/flip.html" class="external-link">terra::flip()</a></code>:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/mirrorImg.html">mirrorImg</a></span><span class="op">(</span><span class="va">img</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/imgRaster.html">imgRaster</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-78-1.png" width="96"></p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/mirrorImg.html">mirrorImg</a></span><span class="op">(</span><span class="va">img</span>, direction <span class="op">=</span> <span class="st">"horizontal"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/imgRaster.html">imgRaster</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-78-2.png" width="96"></p>
<p>Here we apply the transformation to an SFE object, where the image
specified by the sample and image IDs are transformed:</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mirrorImg.html">mirrorImg</a></span><span class="op">(</span><span class="va">sfe3</span>, sample_id <span class="op">=</span> <span class="st">"sample01"</span>, image_id <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span></span></code></pre></div>
<p>So far, <code><a href="../reference/transposeImg.html">transposeImg()</a></code> and <code><a href="../reference/mirrorImg.html">mirrorImg()</a></code> only
transform the image. But the entire SFE object, including all the
geometries and images, can be transformed at once.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_mirrored</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFE-transform.html">mirror</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span></span>
<span><span class="va">sfe_transposed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFE-transform.html">transpose</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_mirrored</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe_mirrored</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotGeometries.html">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_transposed</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/colGeometries.html">spotPoly</a></span><span class="op">(</span><span class="va">sfe_transposed</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SFE_files/figure-html/unnamed-chunk-81-1.png" width="576"></p>
<p>Transforming the entire SFE object can be useful when the tissue has
a orientation and a conventional direction of the orientation, such as
rostral is conventionally at the top while caudal is at the bottom in
coronal brain sections, while anterior is at the left and posterior is
at the right in saggital brain sections, to make data conform to the
convention.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="limitations-and-future-directions">Limitations and future directions<a class="anchor" aria-label="anchor" href="#limitations-and-future-directions"></a>
</h2>
<p>These are the limitations of the current version of SFE:</p>
<ol style="list-style-type: decimal">
<li>By integrating with <code>sf</code>, which is designed for vector
spatial data (specifying coordinates of points, lines, and polygons
vertices), SFE only supports vector data for the geometries, and raster
(like an image, with a value at each pixel) is not supported. Vector is
chosen, as it is a more memory efficient way to represent cell and
nuclei segmentation than a raster map.</li>
<li>The spatial graphs are <code>listw</code> objects so no conversion
is necessary to use the well-established spatial statistical methods in
the <code>spdep</code>, <code>spatialreg</code>, and
<code>adespatial</code> packages. However, <code>igraph</code>
implements many graph analysis methods, and conversion is required to
use them. Whether future versions of SFE will stick to
<code>listw</code> depends on importance of methods that use spatial
graphs in <code>igraph</code> class.</li>
<li>While Simple Features support 3D and spatiotemporal coordinates,
most geospatial resources SFE leverages <code>sf</code> for is for 2D
data.</li>
<li>Spatial point process analysis with the <code>spatstat</code>
package may be relevant, such as in analyzing spatial distribution of
nuclei or transcript spots. As <code>spatstat</code> predates
<code>sf</code> by over a decade, <code>spatstat</code> does not play
very nicely with <code>sf</code>. However, since analyses of nuclei and
transcript spot localization don’t center on the gene count matrix,
whether <code>spatstat</code> analyses should be integrated into SFE
(which is centered on the gene count matrix) is questionable.</li>
<li>Geometries for very large datasets can get very large. On disk
operations of the geometries should be considered. The geospatial field
already has on disk tools for both vector and raster data. So far, SFE
has only been tested on data that fit into memory.</li>
<li>Setting units of length in the SFE object and converting units. This
can make geometries of different samples and datasets comparable, and
helpful to plotting scale bars when plotting geometries.</li>
</ol>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean up</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"vizgen"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"cosmx"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8          LC_NUMERIC=C             </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=C.UTF-8           LC_COLLATE=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=C.UTF-8       LC_MESSAGES=C.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8          LC_NAME=C.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C.UTF-8        LC_TELEPHONE=C.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8    LC_IDENTIFICATION=C.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] RBioFormats_1.5.1              Matrix_1.7-0                  </span></span>
<span><span class="co">#&gt;  [3] sf_1.0-16                      SFEData_1.7.0                 </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.15.1       SingleCellExperiment_1.27.2   </span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.35.1    Biobase_2.65.0                </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.57.1           GenomeInfoDb_1.41.1           </span></span>
<span><span class="co">#&gt; [11] IRanges_2.39.2                 S4Vectors_0.43.2              </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.51.0            MatrixGenerics_1.17.0         </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.3.0              SpatialFeatureExperiment_1.7.1</span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.33.1              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] spatstat.sparse_3.1-0     fs_1.6.4                 </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-7              EBImage_4.47.0           </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.7                RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;   [7] sctransform_0.4.1         tools_4.4.1              </span></span>
<span><span class="co">#&gt;   [9] utf8_1.2.4                R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [11] HDF5Array_1.33.5          uwot_0.2.2               </span></span>
<span><span class="co">#&gt;  [13] lazyeval_0.2.2            rhdf5filters_1.17.0      </span></span>
<span><span class="co">#&gt;  [15] withr_3.0.0               sp_2.1-4                 </span></span>
<span><span class="co">#&gt;  [17] gridExtra_2.3             progressr_0.14.0         </span></span>
<span><span class="co">#&gt;  [19] cli_3.6.3                 textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [21] spatstat.explore_3.3-1    fastDummies_1.7.3        </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.9                Seurat_5.1.0             </span></span>
<span><span class="co">#&gt;  [25] spatstat.data_3.1-2       arrow_16.1.0             </span></span>
<span><span class="co">#&gt;  [27] proxy_0.4-27              ggridges_0.5.6           </span></span>
<span><span class="co">#&gt;  [29] pbapply_1.7-2             pkgdown_2.1.0            </span></span>
<span><span class="co">#&gt;  [31] systemfonts_1.1.0         R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [33] parallelly_1.37.1         limma_3.61.4             </span></span>
<span><span class="co">#&gt;  [35] RSQLite_2.3.7             generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [37] spatstat.random_3.3-1     ica_1.0-3                </span></span>
<span><span class="co">#&gt;  [39] spdep_1.3-5               dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [41] fansi_1.0.6               abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [43] R.methodsS3_1.8.2         terra_1.7-78             </span></span>
<span><span class="co">#&gt;  [45] lifecycle_1.0.4           yaml_2.3.9               </span></span>
<span><span class="co">#&gt;  [47] edgeR_4.3.5               rhdf5_2.49.0             </span></span>
<span><span class="co">#&gt;  [49] SparseArray_1.5.25        BiocFileCache_2.13.0     </span></span>
<span><span class="co">#&gt;  [51] Rtsne_0.17                grid_4.4.1               </span></span>
<span><span class="co">#&gt;  [53] blob_1.2.4                promises_1.3.0           </span></span>
<span><span class="co">#&gt;  [55] dqrng_0.4.1               ExperimentHub_2.13.0     </span></span>
<span><span class="co">#&gt;  [57] crayon_1.5.3              miniUI_0.1.1.1           </span></span>
<span><span class="co">#&gt;  [59] lattice_0.22-6            beachmat_2.21.4          </span></span>
<span><span class="co">#&gt;  [61] cowplot_1.1.3             KEGGREST_1.45.1          </span></span>
<span><span class="co">#&gt;  [63] magick_2.8.4              zeallot_0.1.0            </span></span>
<span><span class="co">#&gt;  [65] pillar_1.9.0              knitr_1.48               </span></span>
<span><span class="co">#&gt;  [67] rjson_0.2.21              boot_1.3-30              </span></span>
<span><span class="co">#&gt;  [69] future.apply_1.11.2       sfarrow_0.4.1            </span></span>
<span><span class="co">#&gt;  [71] codetools_0.2-20          leiden_0.4.3.1           </span></span>
<span><span class="co">#&gt;  [73] wk_0.9.2                  glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [75] spatstat.univar_3.0-0     data.table_1.15.4        </span></span>
<span><span class="co">#&gt;  [77] vctrs_0.6.5               png_0.1-8                </span></span>
<span><span class="co">#&gt;  [79] spam_2.10-0               gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [81] assertthat_0.2.1          cachem_1.1.0             </span></span>
<span><span class="co">#&gt;  [83] xfun_0.46                 S4Arrays_1.5.5           </span></span>
<span><span class="co">#&gt;  [85] mime_0.12                 DropletUtils_1.25.0      </span></span>
<span><span class="co">#&gt;  [87] survival_3.7-0            sfheaders_0.4.4          </span></span>
<span><span class="co">#&gt;  [89] rJava_1.0-11              units_0.8-5              </span></span>
<span><span class="co">#&gt;  [91] statmod_1.5.0             fitdistrplus_1.2-1       </span></span>
<span><span class="co">#&gt;  [93] ROCR_1.0-11               nlme_3.1-165             </span></span>
<span><span class="co">#&gt;  [95] bit64_4.0.5               filelock_1.0.3           </span></span>
<span><span class="co">#&gt;  [97] RcppAnnoy_0.0.22          bslib_0.7.0              </span></span>
<span><span class="co">#&gt;  [99] irlba_2.3.5.1             KernSmooth_2.23-24       </span></span>
<span><span class="co">#&gt; [101] colorspace_2.1-0          spData_2.3.1             </span></span>
<span><span class="co">#&gt; [103] DBI_1.2.3                 tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt; [105] bit_4.0.5                 compiler_4.4.1           </span></span>
<span><span class="co">#&gt; [107] curl_5.2.1                BiocNeighbors_1.23.0     </span></span>
<span><span class="co">#&gt; [109] xml2_1.3.6                desc_1.4.3               </span></span>
<span><span class="co">#&gt; [111] DelayedArray_0.31.9       plotly_4.10.4            </span></span>
<span><span class="co">#&gt; [113] bookdown_0.40             scales_1.3.0             </span></span>
<span><span class="co">#&gt; [115] classInt_0.4-10           lmtest_0.9-40            </span></span>
<span><span class="co">#&gt; [117] rappdirs_0.3.3            goftest_1.2-3            </span></span>
<span><span class="co">#&gt; [119] stringr_1.5.1             tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [121] digest_0.6.36             spatstat.utils_3.0-5     </span></span>
<span><span class="co">#&gt; [123] fftwtools_0.9-11          rmarkdown_2.27           </span></span>
<span><span class="co">#&gt; [125] XVector_0.45.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt; [127] pkgconfig_2.0.3           jpeg_0.1-10              </span></span>
<span><span class="co">#&gt; [129] sparseMatrixStats_1.17.2  highr_0.11               </span></span>
<span><span class="co">#&gt; [131] dbplyr_2.5.0              fastmap_1.2.0            </span></span>
<span><span class="co">#&gt; [133] rlang_1.1.4               htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt; [135] UCSC.utils_1.1.0          shiny_1.8.1.1            </span></span>
<span><span class="co">#&gt; [137] DelayedMatrixStats_1.27.2 jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt; [139] zoo_1.8-12                jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt; [141] BiocParallel_1.39.0       R.oo_1.26.0              </span></span>
<span><span class="co">#&gt; [143] RCurl_1.98-1.16           magrittr_2.0.3           </span></span>
<span><span class="co">#&gt; [145] scuttle_1.15.2            GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt; [147] s2_1.1.7                  dotCall64_1.1-1          </span></span>
<span><span class="co">#&gt; [149] patchwork_1.2.0           Rhdf5lib_1.27.0          </span></span>
<span><span class="co">#&gt; [151] munsell_0.5.1             Rcpp_1.0.13              </span></span>
<span><span class="co">#&gt; [153] reticulate_1.38.0         stringi_1.8.4            </span></span>
<span><span class="co">#&gt; [155] zlibbioc_1.51.1           MASS_7.3-61              </span></span>
<span><span class="co">#&gt; [157] plyr_1.8.9                AnnotationHub_3.13.0     </span></span>
<span><span class="co">#&gt; [159] parallel_4.4.1            listenv_0.9.1            </span></span>
<span><span class="co">#&gt; [161] ggrepel_0.9.5             deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [163] Biostrings_2.73.1         splines_4.4.1            </span></span>
<span><span class="co">#&gt; [165] tensor_1.5                locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt; [167] igraph_2.0.3              spatstat.geom_3.3-2      </span></span>
<span><span class="co">#&gt; [169] RcppHNSW_0.6.0            reshape2_1.4.4           </span></span>
<span><span class="co">#&gt; [171] BiocVersion_3.20.0        evaluate_0.24.0          </span></span>
<span><span class="co">#&gt; [173] SeuratObject_5.0.2        BiocManager_1.30.23      </span></span>
<span><span class="co">#&gt; [175] httpuv_1.6.15             polyclip_1.10-6          </span></span>
<span><span class="co">#&gt; [177] RANN_2.6.1                tidyr_1.3.1              </span></span>
<span><span class="co">#&gt; [179] purrr_1.0.2               future_1.33.2            </span></span>
<span><span class="co">#&gt; [181] scattermore_1.2           ggplot2_3.5.1            </span></span>
<span><span class="co">#&gt; [183] xtable_1.8-4              e1071_1.7-14             </span></span>
<span><span class="co">#&gt; [185] RSpectra_0.16-2           later_1.3.2              </span></span>
<span><span class="co">#&gt; [187] viridisLite_0.4.2         class_7.3-22             </span></span>
<span><span class="co">#&gt; [189] ragg_1.3.2                tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [191] memoise_2.0.1             AnnotationDbi_1.67.0     </span></span>
<span><span class="co">#&gt; [193] cluster_2.1.6             globals_0.16.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
